server: TestServerShutdownReleasesSession failed

server.TestServerShutdownReleasesSession [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/11060609?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/11060609?buildTab=artifacts#/) on master @ [640a56415d3aa263b85d29fcd0089a43f2f7e95a](https://github.com/cockroachdb/cockroach/commits/640a56415d3aa263b85d29fcd0089a43f2f7e95a):


```
=== RUN   TestServerShutdownReleasesSession
    test_log_scope.go:167: test logs captured to: /artifacts/tmp/_tmp/7f3306bd25c63f593c1cefe649bebfca/logTestServerShutdownReleasesSession4183567719
    test_log_scope.go:81: use -show-logs to present logs inline
    drain_test.go:336: error executing 'SELECT session_id FROM system.sqlliveness WHERE session_id = $1': pq: failed to connect to n2 at : initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp: missing address" [code 14/Unavailable]
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/7f3306bd25c63f593c1cefe649bebfca/logTestServerShutdownReleasesSession4183567719
--- FAIL: TestServerShutdownReleasesSession (6.43s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv @cockroachdb/server
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestServerShutdownReleasesSession.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-30119

More details on the failure that happened:

After calling `DrainClients`, the test executes this query explicitly: `SELECT session_id FROM system.sqlliveness WHERE session_id = $1`

The query failed with an error from the grpc layer: `pq: failed to connect to n2 at : initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp: missing address" [code 14/Unavailable]`.

Here are some related server logs:
```
I230726 02:46:45.284098 2523 server/drain.go:451 â‹® [-] 295  SQL server drained successfully; SQL queries cannot execute any more
W230726 02:46:45.284128 7958 sql/sqlliveness/slinstance/slinstance.go:334 â‹® [T10,nsql2] 296  exiting heartbeat loop
I230726 02:46:45.284137 7958 sql/sqlliveness/slinstance/slinstance.go:319 â‹® [T10,nsql2] 297  draining heartbeat loop
E230726 02:46:45.292188 8465 2@rpc/peer.go:611 â‹® [T10,nsql1,rnode=2,raddr,class=default,rpc] 298  failed connection attemptâ€¹ (last connected 0s ago)â€º: grpc: â€¹connection error: desc = "transport: error while dialing: dial tcp: missing address"â€º [code 14/Unavailable]
E230726 02:46:45.293316 8548 2@rpc/peer.go:611 â‹® [T10,nsql1,rnode=2,raddr,class=default,rpc] 299  failed connection attemptâ€¹ (last connected 1ms ago)â€º: grpc: â€¹connection error: desc = "transport: error while dialing: dial tcp: missing address"â€º [code 14/Unavailable]
I230726 02:46:45.294127 5786 sql/stats/automatic_stats.go:572 â‹® [T10,nsql1] 300  quiescing auto stats refresher
I230726 02:46:45.294198 5787 sql/stats/automatic_stats.go:624 â‹® [T10,nsql1] 301  quiescing stats garbage collector
I230726 02:46:45.294321 5667 server/start_listen.go:103 â‹® [T10,nsql1] 302  server shutting down: instructing cmux to stop accepting
W230726 02:46:45.294510 5671 kv/kvclient/kvtenant/connector.go:387 â‹® [T10,nsql1,tenant-connector] 303  error consuming GossipSubscription RPC: recv msg error: grpc: â€¹context canceledâ€º [code 1/Canceled]
I230726 02:46:45.294749 5989 jobs/registry.go:1606 â‹® [T10,nsql1] 304  AUTO SPAN CONFIG RECONCILIATION job 885618127142125569: stepping through state succeeded
W230726 02:46:45.296828 5750 sql/sqlliveness/slinstance/slinstance.go:334 â‹® [T10,nsql1] 305  exiting heartbeat loop
W230726 02:46:45.296852 5750 sql/sqlliveness/slinstance/slinstance.go:321 â‹® [T10,nsql1] 306  exiting heartbeat loop with error: node unavailable; try another peer
W230726 02:46:45.296932 5672 kv/kvclient/kvtenant/setting_overrides.go:59 â‹® [T10,nsql1,tenant-connector] 307  error consuming TenantSettings RPC: recv msg error: grpc: â€¹context canceledâ€º [code 1/Canceled]
E230726 02:46:45.297048 7152 jobs/registry.go:1004 â‹® [-] 308  error getting live session: node unavailable; try another peer
I230726 02:46:45.298703 8553 server/node.go:1341 â‹® [T1,n1,tenant=10] 309  batch request â€¹Scan [/Tenant/10/Table/15/1/885618127142125569,/Tenant/10/Table/15/1/885618127142125570), [txn: 8a3e7723], [can-forward-ts]â€º failed with error: aborted during Replica.Send: context canceled
I230726 02:46:45.298703 8553 server/node.go:1341 â‹® [T1,n1,tenant=10] 309 +trace:
I230726 02:46:45.298703 8553 server/node.go:1341 â‹® [T1,n1,tenant=10] 309 +     0.000ms      0.000ms    === operation:/cockroach.roachpb.Internal/Batch
W230726 02:46:45.301934 5989 jobs/adopt.go:534 â‹® [-] 310  could not clear job claim: â€¹clear-job-claim: failed to send RPC: sending to all replicas failed; last error: ba: Scan [/Tenant/10/Table/15/1/885618127142125569,/Tenant/10/Table/15/1/885618127142125570), [txn: 8a3e7723], [can-forward-ts] RPC error: grpc: error reading from server: read tcp 127.0.0.1:57440->127.0.0.1:38771: use of closed network connection [code 14/Unavailable]â€º
W230726 02:46:45.306597 2523 server/server_sql.go:1676 â‹® [T10,nsql1] 311  server shutdown without a prior graceful drain
I230726 02:46:45.306822 3423 sql/stats/automatic_stats.go:624 â‹® [T1,n1] 312  quiescing stats garbage collector
I230726 02:46:45.307056 2648 server/start_listen.go:103 â‹® [T1,n1] 313  server shutting down: instructing cmux to stop accepting
W230726 02:46:45.307122 7537 kv/kvserver/intentresolver/intent_resolver.go:826 â‹® [-] 314  failed to gc transaction record: could not GC completed transaction anchored at /Tenant/10/Table/15/1/â€¹885618130467422209â€º: node unavailable; try another peer
```

server.TestServerShutdownReleasesSession [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/11102955?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/11102955?buildTab=artifacts#/) on master @ [167da6563fbc62ff7ad447bcf44bc24f336aef6b](https://github.com/cockroachdb/cockroach/commits/167da6563fbc62ff7ad447bcf44bc24f336aef6b):


```
=== RUN   TestServerShutdownReleasesSession
    test_log_scope.go:167: test logs captured to: /artifacts/tmp/_tmp/7f3306bd25c63f593c1cefe649bebfca/logTestServerShutdownReleasesSession3453402071
    test_log_scope.go:81: use -show-logs to present logs inline
    drain_test.go:336: error executing 'SELECT session_id FROM system.sqlliveness WHERE session_id = $1': pq: failed to connect to n2 at : initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp: missing address" [code 14/Unavailable]
    panic.go:522: -- test log scope end --
--- FAIL: TestServerShutdownReleasesSession (4.26s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestServerShutdownReleasesSession.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/server/drain_test.go
+++ b/pkg/server/drain_test.go
@@ -346,6 +346,7 @@ func TestServerShutdownReleasesSession(t *testing.T) {
 	require.True(t, sessionExists(*session))
 
 	require.NoError(t, tmpTenant.DrainClients(context.Background()))
+	tmpTenant.Stopper().Stop(ctx)
 
 	require.False(t, sessionExists(*session), "expected session %s to be deleted from the sqlliveness table, but it still exists", *session)
 	require.Nil(t, queryOwner(tmpSQLInstance), "expected sql_instance %d to have no owning session_id", tmpSQLInstance)

